src/operator/nn/pool_utils.h

```diff
@@ -0,0 +1,115 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+#ifndef MXNET_OPERATOR_NN_POOL_UTILS_H_
+#define MXNET_OPERATOR_NN_POOL_UTILS_H_
+
+#include "../mshadow_op.h"
+
+namespace mxnet {
+namespace op {
+
+template<typename DType, int p>
+struct a_to_p {
+  static MSHADOW_XINLINE DType Map(DType a) {
+    return mshadow_op::power::Map(a, DType(p));
+  }
+};
+
+template<typename DType>
+struct a_to_p<DType, 1> {
+  static MSHADOW_XINLINE DType Map(DType a) {
+    return a;
+  }
+};
+
+template<typename DType>
+struct a_to_p<DType, 2> {
+  static MSHADOW_XINLINE DType Map(DType a) {
+    return a*a;
+  }
+};
+
+template<typename DType>
+struct a_to_p<DType, 3> {
+  static MSHADOW_XINLINE DType Map(DType a) {
+    return a*a*a;
+  }
+};
+
+template<typename DType, int p>
+struct a_root_p {
+  static MSHADOW_XINLINE DType Map(DType a) {
+    return mshadow_op::power::Map(a, DType(1.0 / p));
+  }
+};
+
+template<typename DType>
+struct a_root_p<DType, 1> {
+  static MSHADOW_XINLINE DType Map(DType a) {
+    return a;
+  }
+};
+
+template<typename DType>
+struct a_root_p<DType, 2> {
+  static MSHADOW_XINLINE DType Map(DType a) {
+    return mshadow_op::square_root::Map(a);
+  }
+};
+
+template<typename DType>
+struct a_root_p<DType, 3> {
+  static MSHADOW_XINLINE DType Map(DType a) {
+    return mshadow_op::cube_root::Map(a);
+  }
+};
+
+template<typename DType, int p>
+struct lp_grad {
+  static MSHADOW_XINLINE DType Map(DType grad, DType in_data, DType out_data) {
+    return grad * mshadow_op::power::Map(in_data / out_data, DType(p - 1));
+  }
+};
+
+template<typename DType>
+struct lp_grad<DType, 1> {
+  static MSHADOW_XINLINE DType Map(DType grad, DType in_data, DType out_data) {
+    return grad;
+  }
+};
+
+template<typename DType>
+struct lp_grad<DType, 2> {
+  static MSHADOW_XINLINE DType Map(DType grad, DType in_data, DType out_data) {
+    return grad * in_data / out_data;
+  }
+};
+
+template<typename DType>
+struct lp_grad<DType, 3> {
+  static MSHADOW_XINLINE DType Map(DType grad, DType in_data, DType out_data) {
+    return grad * in_data * in_data / out_data / out_data;
```

Comment:
Floating number division is slow. You can reduce 2 divisions into 1 and it would be faster.

Reply:
Do you mean something like ```return grad * in_data * in_data / (out_data * out_data)```?